aspnetcore/fundamentals/servers/kestrel.md

```diff
@@ -0,0 +1,135 @@
+
+## How to use Kestrel in ASP.NET Core apps
+
+Install the [Microsoft.AspNetCore.Server.Kestrel](https://www.nuget.org/packages/Microsoft.AspNetCore.Server.Kestrel/) NuGet package.
+
+Call the [`UseKestrel`](http://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNetCore/Hosting/WebHostBuilderKestrelExtensions/index.html#Microsoft.AspNetCore.Hosting.WebHostBuilderKestrelExtensions.UseKestrel.md) extension method on `WebHostBuilder` in your `Main` method, specifying any [Kestrel options](https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNetCore/Server/Kestrel/KestrelServerOptions/) that you need, as shown in the following example:
+
+[!code-csharp[](kestrel/sample/Program.cs?name=snippet_Main&highlight=13-19)]
+
+### URL prefixes
+
+By default Kestrel binds to `http://localhost:5000`. You can configure URL prefixes and ports for Kestrel to listen on by using the `UseURLs` extension method, the `urls` command-line argument, or the ASP.NET Core configuration system. For more information about these methods, see [Hosting](../../fundamentals/hosting.md). For information about how URL binding works when you use IIS as a reverse proxy, see [ASP.NET Core Module](aspnet-core-module.md). 
+
+URL prefixes for Kestrel can be in any of the following formats. 
+
+* IPv4 address with port number
+
+  ```
+  http://65.55.39.10:80/
+  https://65.55.39.10:443/
+  ```
+
+  Address 0.0.0.0 is a special case that binds to all IPv4 addresses.
+
+
+* IPv6 address with port number
+
+  ```
+  http://[0:0:0:0:0:ffff:4137:270a]:80/ 
+  https://[0:0:0:0:0:ffff:4137:270a]:443/ 
+  ```
+
+* Host name with port number
+
+  ```
+  http://contoso.com:80/
+  http://*:80/
+  https://contoso.com:443/
+  https://*:443/
+  ````
+
+  Host names, *, and +, are not special. Anything that is not a recognized IP address or "localhost" will bind to all IPv4 and IPv6 IPs. If you need to bind different host names to different ASP.NET Core applications on the same port, use [WebListener](weblistener.md) or a reverse proxy server such as IIS, Nginx, or Apache.
+
+* "Localhost" name with port number or loopback IP with port number
+
+  ```
+  http://localhost:5000/
+  http://127.0.0.1:5000/
+  http://[::1]:5000/
+  ```
+
+  When `localhost` is specified, Kestrel tries to bind to both IPv4 and IPv6 loopback interfaces. If the requested port is in use by another service on either loopback interface, Kestrel fails to start. If either loopback interface is unavailable for any other reason (most commonly because IPv6 is not supported), Kestrel logs a warning. 
+
+* Unix socket
+
+  ```
+  http://unix:/run/dan-live.sock  
+  ```
+
+If you specify port number 0, Kestrel dynamically binds to an available port. Binding to port 0 is allowed for any host name or IP except for loopback IPs.
```

Comment:
Loopback IP addresses are fine for port 0. We only disallow dynamic binding to `localhost`.